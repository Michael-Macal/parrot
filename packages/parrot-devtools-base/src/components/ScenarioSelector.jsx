import React, { PropTypes } from 'react';
import { Dropdown } from 'axp-base';

export const ScenarioSelector = (props) => {
  const onScenarioChange = (newVal) => {
    fetch('/scenario', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        scenario: newVal.value
      })
    })
    .catch((err) => { console.warn(err); })
    .then(() => { props.refetch(); });
  };

  const options = Object.keys(props.scenarios).map(scenario => ({
    value: scenario,
    text: scenario
  }));

  return (
    <Dropdown
      id="scenario-select"
      options={ options }
      label="Scenario"
      onChange={ onScenarioChange }
      value={ props.selected }
    />
  );
};

ScenarioSelector.propTypes = {
  selected: PropTypes.string,
  scenarios: PropTypes.object,
  refetch: PropTypes.func
};

ScenarioSelector.defaultProps = {
  refetch: () => location.reload()
};

export default ScenarioSelector;
