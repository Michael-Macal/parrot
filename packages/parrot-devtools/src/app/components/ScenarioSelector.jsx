import React, { Component, PropTypes } from 'react';
import url from 'url';
import SelectField from 'material-ui/SelectField';
import MenuItem from 'material-ui/MenuItem';
import CircularProgress from 'material-ui/CircularProgress';

class ScenarioSelector extends Component {
  static propTypes = {
    url: PropTypes.string.isRequired, // eslint-disable-line react/no-unused-prop-types
  }

  state = {
    loading: true,
    scenarios: {},
    scenario: '',
    error: null,
  };

  async componentDidMount() {
    try {
      const resp = await fetch(this.createUrl('/scenarios'), {
        headers: {
          Accept: 'application/json',
        },
      });

      const scenarios = await resp.json();
      const keys = Object.keys(scenarios);

      this.setState({ // eslint-disable-line react/no-did-mount-set-state
        loading: false,
        scenarios,
        scenario: keys[0],
      });
    } catch (error) {
      this.setState({ error }); // eslint-disable-line react/no-did-mount-set-state
    }
  }

  setScenario = (event, key, scenario) => {
    this.setState({
      scenario,
    }, async () => {
      try {
        const resp = await fetch(this.createUrl('/scenario'), {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            scenario,
          }),
        });

        if (!resp.ok) {
          this.setState({ error: resp.statusText });
        } else if (chrome) { // Reload if we are in chrome
          chrome.devtools.inspectedWindow.reload(null);
        }
      } catch ({ message: error }) {
        this.setState({ error });
      }
    });
  }

  createUrl = pathname => url.format({ ...url.parse(this.props.url), pathname })

  render() {
    if (this.state.error) {
      return <pre>{this.state.error}</pre>;
    } else if (this.state.loading) {
      return <CircularProgress />;
    }

    return (
      <SelectField
        floatingLabelText="Scenario"
        value={this.state.scenario}
        onChange={this.setScenario}
      >
        {Object.keys(this.state.scenarios).map(scenario =>
          <MenuItem key={scenario} value={scenario} primaryText={scenario} />,
        )}
      </SelectField>
    );
  }
}

export default ScenarioSelector;
