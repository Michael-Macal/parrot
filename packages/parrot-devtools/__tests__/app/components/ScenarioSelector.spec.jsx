/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React from 'react';
import { shallow } from 'enzyme';
import ScenarioSelector from '../../../src/app/components/ScenarioSelector';

describe('ScenarioSelector', () => {
  it('should render loading', () => {
    expect(
      shallow(
        <ScenarioSelector
          data={{ setScenario: () => null, loadScenarios: () => null }}
          loading={true}
        />
      )
    ).toMatchSnapshot();
  });

  it('should render error', () => {
    expect(
      shallow(
        <ScenarioSelector
          data={{ scenarios: [], setScenario: () => null, loadScenarios: () => null }}
          loading={false}
        />
      )
    ).toMatchSnapshot();
  });

  it('should render scenarios', () => {
    expect(
      shallow(
        <ScenarioSelector
          data={{
            scenarios: [{ name: 'squawk' }],
            scenario: 'squawk',
            setScenario: () => null,
            loadScenarios: () => null,
          }}
          loading={false}
        />
      )
    ).toMatchSnapshot();
  });

  it('should setScenario', () => {
    const setScenario = jest.fn();
    const rendered = shallow(
      <ScenarioSelector
        data={{
          scenarios: [{ name: 'squawk' }],
          scenario: 'squawk',
          setScenario,
          loadScenarios: () => null,
        }}
        loading={false}
      />
    );

    rendered.find('a').simulate('click');
    expect(setScenario).toHaveBeenCalledWith('squawk');
  });
});
