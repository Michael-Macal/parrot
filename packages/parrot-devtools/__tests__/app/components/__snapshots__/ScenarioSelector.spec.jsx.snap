// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScenarioSelector should render error 1`] = `
<styled.div>
  <div
    className="alert alert-warn alert-dialog alert-lg"
    role="alert"
  >
    <span
      className="alert-icon dls-icon-warning icon-lg"
    />
    <h1
      className="heading-4 alert-header"
    >
      Failed to Load Scenarios
    </h1>
    <p
      className="alert-subtitle"
    >
      Try 
      <a
        onClick={[Function]}
      >
        refreshing
      </a>
       or updating the middleware URL in settings.
    </p>
  </div>
</styled.div>
`;

exports[`ScenarioSelector should render loading 1`] = `
<styled.div>
  <div
    className="progress-circle progress-indeterminate margin-auto"
  />
</styled.div>
`;

exports[`ScenarioSelector should render scenarios 1`] = `
<styled.div
  className="border"
>
  <ul
    className="nav-menu"
  >
    <li
      className="nav-item"
      key="squawk"
    >
      <a
        aria-selected={true}
        className="nav-link fluid"
        onClick={[Function]}
      >
        squawk
      </a>
    </li>
  </ul>
</styled.div>
`;
